### ISSUE TITLE
Permit Opaque types as fields of shared structs, making the shared struct Opaque

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/cxx#340]
```rust
#[cxx::bridge]
mod ffi {
    struct S {
        s: CxxString,
        t: T,
    }

    extern "C" {
        type T;

        fn f() -> UniquePtr<S>;
        fn g(s: &S);
    }
}
```

The above should be sound to allow. Other things in the library design already enforce that Rust code is unable to obtain a CxxString or T by value, and as such it would be impossible to improperly construct an S by value in Rust.

@adetaylor: I imagine this is very relevant to https://github.com/google/autocxx since lots of things probably use C++ strings as fields.

**Comment by dtolnay** [comment]
The current output:

```console
error: using C++ string by value is not supported
 --> demo/src/main.rs:4:9
  |
4 |         s: CxxString,
  |         ^^^^^^^^^^^^

error: using opaque C++ type by value is not supported
 --> demo/src/main.rs:5:9
  |
5 |         t: T,
  |         ^^^^
```

**Comment by adetaylor** [comment]
Yeah. `autocxx` at the moment currently generates lots of shared structs, but I'm soon going to merge a branch which will switch us over to solely extern types (which is what I should have used in the first place!)

**Comment by dtolnay** [comment]
Note if we allow this and a struct has an opaque type field that is not the last field, we'd need to either produce accessor methods or one of the sneaky approaches from https://github.com/google/autocxx/issues/21. ([playground1](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8d66743adb1841497ec14d6809b68092), [playground2](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=be739d2067d0b205819e317262843fa9))
